### PR TITLE
Test additional NativeAOT Lib testing

### PR DESCRIPTION
**Pull request by LakshanF**
First take on setting up a NativeAOT additional library run on `Runtime-extra-platforms` pipeline

### CODE REVIEW

File: eng/pipelines/runtime-extra-platforms-other.yml [hunk 1]
```diff
@@ -72,6 +72,35 @@ jobs:
...
+# CoreCLR NativeAOT release build and additional libraries tests
+# Only when CoreCLR or library is changed
+#
+- template: /eng/pipelines/common/platform-matrix.yml
+  parameters:
+    jobTemplate: /eng/pipelines/common/global-build-job.yml
+    helixQueuesTemplate: /eng/pipelines/libraries/helix-queues-setup.yml
+    buildConfig: Release
+    platforms:
+    - windows_x64
```
Comment (Beau-Gosse-dev) [inline eng/pipelines/runtime-extra-platforms-other.yml:85]: Just a question for my own understanding, could you have done Windows ARM here too, like in the [runtime.yml](https://github.com/dotnet/runtime/blob/218a4b1dad3f73eec8f9898761501c4cfa14698e/eng/pipelines/runtime.yml#L257), but we just chose x64 to get different coverage?

File: eng/pipelines/runtime-extra-platforms-other.yml [hunk 2]
```diff
@@ -72,6 +72,35 @@ jobs:
...
+      testGroup: innerloop
+      isSingleFile: true
+      nameSuffix: NativeAOT_Additional_Libs
+      buildArgs: -s clr.alljits+clr.tools+clr.nativeaotlibs+clr.nativeaotruntime+libs+libs.tests -c $(_BuildConfig) /p:TestNativeAot=true /p:TestNativeAotAdditionalLibs=true /p:ArchiveTests=true
+      timeoutInMinutes: 180
+      # extra steps, run tests
+      extraStepsTemplate: /eng/pipelines/libraries/helix.yml
+      extraStepsParameters:
+        creator: dotnet-bot
+        testRunNamePrefixSuffix: NativeAOT_$(_BuildConfig)
```
Comment (Beau-Gosse-dev) [inline eng/pipelines/runtime-extra-platforms-other.yml:97]: Would you want the testRunNamePrefixSuffix to also be "NativeAOT_Additional_Libs" or will the nameSuffix already show that?

File: eng/pipelines/runtime-extra-platforms-other.yml [hunk 3]
```diff
@@ -72,6 +72,35 @@ jobs:
...
+    helixQueuesTemplate: /eng/pipelines/libraries/helix-queues-setup.yml
+    buildConfig: Release
+    platforms:
+    - windows_x64
+    - windows_arm64
+    jobParameters:
+      testGroup: innerloop
+      isSingleFile: true
+      nameSuffix: NativeAOT_Additional_Libs
+      buildArgs: -s clr.alljits+clr.tools+clr.nativeaotlibs+clr.nativeaotruntime+libs+libs.tests -c $(_BuildConfig) /p:TestNativeAot=true /p:TestNativeAotAdditionalLibs=true /p:ArchiveTests=true
```
Comment (Beau-Gosse-dev) [inline eng/pipelines/runtime-extra-platforms-other.yml:91]: Is it standard to use both build parameters "TestNativeAot" and "TestNativeAotAdditionalLibs". I would have thought we'd just use the new one. Either way is fine with me, I'm just trying to learn what the norm is.

**Review by Beau-Gosse-dev — COMMENTED** [review]
Looks good to me, just have a few questions to help me better understand some stuff.

File: eng/pipelines/runtime-extra-platforms-other.yml (hunk 3 above)
Comment (LakshanF) [reply eng/pipelines/runtime-extra-platforms-other.yml:91]: I wanted the original property to make it clear that this is NativeAOT testing and the new one to provide additional information. But perhaps using 1 might be simpler

File: eng/pipelines/runtime-extra-platforms-other.yml (hunk 2 above)
Comment (LakshanF) [reply eng/pipelines/runtime-extra-platforms-other.yml:97]: Not sure, hopefully the new pipeline run will make it clearer.

File: eng/pipelines/runtime-extra-platforms-other.yml (hunk 1 above)
Comment (LakshanF) [reply eng/pipelines/runtime-extra-platforms-other.yml:85]: I wanted to restrict the initial platforms to windows given the current GC issues in Linux

File: eng/pipelines/runtime-extra-platforms-other.yml (hunk 2 above)
Comment (LakshanF) [reply eng/pipelines/runtime-extra-platforms-other.yml:97]: The extra-platform pipeline completed and the Job name description, "Build windows x64 Release NativeAOT_Additional_Libs", seems sufficient to me. Let me know if you want testrunname to be changed as well.

**Comment by MichalStrehovsky** [comment]
The Mono team introduced `RunSmokeTestsOnly` property in tests.proj (https://github.com/dotnet/runtime/pull/60727). I wonder if we can reuse that (of course we would have a different definition of what we consider a smoke test in TestNativeAOT). They also have properties such as `RunDisabledWasmTests`. I wonder if we should try to align with that instead of introducing `TestNativeAotAdditionalLibs`.

* `/p:TestNativeAOT+/p:RunSmokeTestsOnly` would run the tests we currently run in the CI
* `/p:TestNativeAOT` would run all tests that are not currently failing.
* `/p:TestNativeAOT+/p:RunDisabledNativeAotTests` would run all the tests, including those that we know are failing and we haven't looked at in detail yet.

With this setup, newly introduced test assemblies automatically start running in rolling builds instead of us having to go and add them to the list of working tests.

The Mono team defines their definition of smoke tests here:

https://github.com/dotnet/runtime/blob/12c31dcc0a069b5734f106f88d42cfdc06736ff3/src/libraries/tests.proj#L413-L414

This is how they do the exclusions. We would have a similar exclusion when RunDisabledNativeAotTests is not true:

https://github.com/dotnet/runtime/blob/12c31dcc0a069b5734f106f88d42cfdc06736ff3/src/libraries/tests.proj#L316-L344

File: src/libraries/tests.proj [hunk 4]
```diff
@@ -388,7 +388,7 @@
     <ProjectExclusions Remove="$(MSBuildThisFileDirectory)System.IO.IsolatedStorage\tests\System.IO.IsolatedStorage.Tests.csproj" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TestNativeAot)' == 'true'">
+  <ItemGroup Condition="'$(TestNativeAot)' == 'true' and '$(RunNativeAotSmokeTestsOnly)' == 'true'">
```
Comment (MichalStrehovsky) [inline src/libraries/tests.proj:391]: Could you make this follow the pattern that Mono team created?

* The existing switch name is `RunSmokeTestsOnly`. We can infer "NativeAOT smoke tests" from this and RunNativeAotTests whenever that's needed. It's probably only needed in one place.
* When `RunSmokeTestsOnly` is true, we start with an empty list of tests. We should add ours around the spot where Mono is doing it (https://github.com/dotnet/runtime/blob/aafe4d465f97fa3a0ae61b18c2887cf84085d25d/src/libraries/tests.proj#L410-L416). Maybe put the existing ItemGroup under `!RunNativeAotTests` and add a new one for `RunnativeAotTests==true`

**Review by MichalStrehovsky — COMMENTED** [review]
The summary of the comments I put in is basically:

1. Forget how TestNativeAot is currently handled in tests.proj.
2. Look at how `RunSmokeTestsOnly` and, for example, `RunDisabledAndroidTests` is set up.
3. Make `TestNativeAot=true`, and `TestNativeAot=true && RunSmokeTestsOnly=true`, and `TestNativeAot=true && RunDisabledNativeAotTests=true` follow that.

It's not great if the strategy behind these switches is different between products because it leads to hard to maintain build system.

**Comment by MichalStrehovsky** [comment]
I pushed out two commits to your branch to unify things:

* The existing RunDisabledXXX pattern is to run all the tests, including those that are disabled. Not just the disabled ones. Might be a poor naming choice, but it's a naming choice consistent with half a dozen other places in this file.
* Made things take the existing paths within the project file instead of excluding existing paths for NativeAOT and duplicating things for NativeAOT.
* (In the second commit) Created a more-or-less unified way to define what is a smoke test for various scenarios we have (the Mono's existing scenario + what we have for NativeAOT).

Please have a look.

**Comment by MichalStrehovsky** [comment]
@LakshanF you probably noticed but this is failing because `System.Runtime.InteropServices\tests\System.Runtime.InteropServices.UnitTests\System.Runtime.InteropServices.Tests.csproj` doesn't compile. The exclusion list instead has `System.Runtime.InteropServices\tests\System.Runtime.InteropServices.Tests.csproj` that doesn't actually exist.

Would it be possible to spin off the SingleFileTestRunner.cs change into a separate pull request while this is worked on? CI being green even though tests are failing is a pretty big problem and I would prefer it to be fixed ASAP while this PR is being iterated on.

**Comment by LakshanF** [comment]
> Would it be possible to spin off the SingleFileTestRunner.cs change into a separate pull request while this is worked on? CI being green even though tests are failing is a pretty big problem and I would prefer it to be fixed ASAP while this PR is being iterated on.

#70238 has the change for the CI test run fix

**Comment by MichalStrehovsky** [comment]
Doing a force push without pulling before erased the changes I pushed before. Could you make it so that this behaves the same as for mono legs. GitHub ui lets you see the diff of the force push.

**Comment by LakshanF** [comment]
Did not analyze the failures fully (did not take a look at non-NativeAOT failures) but took a quick look at the Windows (x64 and ARM64) results

- x64 has 4 runs (8.1, 10, 11, Nano): All have about 155 libraries being run with 4-6 library tests failing. One such run,
https://helix.dot.net/api/jobs/69e6476b-2597-447f-a833-9565d67f9a06/workitems?api-version=2019-06-17
```
155 tests (did not count helix job), failing tests
System.Reflection.InvokeEmit.Tests
System.Reflection.InvokeInterpreted.Tests
System.Runtime.ReflectionInvokeEmit.Tests
System.Runtime.ReflectionInvokeInterpreted.Tests

```
- ARM64 run: 155 assembly tests, 4 failing
https://helix.dot.net/api/jobs/11a4dfe1-7514-4d8f-ab60-e256a036c941/workitems?api-version=2019-06-17
```
155 tests (did not count helix job), Failing
System.Reflection.InvokeEmit.Tests
System.Reflection.InvokeInterpreted.Tests
System.Runtime.ReflectionInvokeEmit.Tests
System.Runtime.ReflectionInvokeInterpreted.Tests

```

File: src/libraries/tests.proj [hunk 5]
```diff
@@ -170,6 +170,7 @@
 
     <!-- https://github.com/dotnet/runtime/issues/50926 -->
     <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Diagnostics.Tracing\tests\System.Diagnostics.Tracing.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj" />
```
Comment (MichalStrehovsky) [inline src/libraries/tests.proj:173]: I would leave this up to the Mono team to disable. They probably want a bug number.

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

**Pull request merged**